tests/test_extension.py

```diff
@@ -0,0 +1,683 @@
+            _scope = ExtensionScope.FUNCTION
+
+        # Follow line should be executed from HttpTrigger/__init__.py script
+        # Instantiate a new function extension base
+        FuncExtensionBase()
+        self.assertEqual(len(self._instance._func_exts), 0)
+
+    def test_app_extension_instantiation_should_throw_error(self):
+        """Application extension is operating on a class level, shouldn't be
+        instantiate by trigger script
+        """
+        class NewAppExtension(metaclass=self._instance):
+            _scope = ExtensionScope.APPLICATION
+            _executed = False
+
+            @staticmethod
+            def init():
+                pass
+
+            @staticmethod
+            def post_function_load_app_level():
+                NewAppExtension.executed = True
+
+        with self.assertRaises(FunctionExtensionException):
+            NewAppExtension()
+
+    def test_invalid_scope_extension_instantiation_should_throw_error(self):
+        """If the _scope is not defined in an extension, it is most likely
+        an invalid extension
+        """
+        class InvalidExtension(metaclass=self._instance):
+            pass
+
+        with self.assertRaises(FunctionExtensionException):
+            InvalidExtension()
+
+    def test_get_function_hooks(self):
+        """If a specific extension is registered in to a function, it should
+        be able to retrieve it
+        """
+        extension_hook = self._instance._func_exts.setdefault(
+            'httptrigger', FuncExtensionHooks(
+                post_function_load=[],
+                pre_invocation=[],
+                post_invocation=[]
+            )
+        )
+
+        extension_hook.post_function_load.append(
+            ExtensionHookMeta(ext_impl=lambda: 'hello', ext_name='world')
+        )
+
+        hook = self._instance.get_function_hooks('HttpTrigger')
+        self.assertEqual(hook.post_function_load[0].ext_impl(), 'hello')
+        self.assertEqual(hook.post_function_load[0].ext_name, 'world')
+
+    def test_get_application_hooks(self):
+        """The application extension should be stored in self._app_hooks
+        """
+        hook_obj = AppExtensionHooks(
+            post_function_load_app_level=[],
+            pre_invocation_app_level=[],
+            post_invocation_app_level=[]
+        )
+        self._instance._app_exts = hook_obj
+        hooks = self._instance.get_application_hooks()
+        self.assertEqual(id(hook_obj), id(hooks))
+
+    def test_get_registered_extensions_json_empty(self):
+        """Ensure the get extension json will return empty if there's nothing
+        registered"""
+        info_json = self._instance.get_registered_extensions_json()
+        self.assertEqual(info_json, r'{}')
+
+    def test_get_registered_extensions_json_function_ext(self):
+        """Ensure the get extension json will return function ext info"""
+        class NewFuncExtension(metaclass=self._instance):
```

Comment:
I see `NewFuncExtension` and `NewAppExtension` being defined multiple times. Is there a way to simplify this duplicate code?

Reply:
Yeah, this is a good catch. Refactored the generation process into _generate_new_func_extension_class and _generate_new_app_extension_class.